src/main.cpp

```diff
@@ -695,40 +732,64 @@ int main(int argc, const char* argv[]) {
-                benchmark.run(ast);
-              }
+                logger->info("Running LLVM backend code generator");
+                CodegenLLVMVisitor visitor(modfile,
+                                           output_dir,
+                                           platform,
+                                           llvm_opt_level,
+                                           !llvm_no_debug,
+                                           llvm_fast_math_flags);
+                visitor.visit_program(*ast);
+                ast_to_nmodl(*ast, filepath("llvm", "mod"));
+                ast_to_json(*ast, filepath("llvm", "json"));
+
+                if (llvm_benchmark) {
+                    logger->info("Running LLVM benchmark");
+                    if (platform.is_gpu() && !platform.is_CUDA_gpu()) {
+                        throw std::runtime_error(
+                            "Benchmarking is only supported on CUDA GPUs at the moment");
+                    }
+                    if (platform.is_gpu()) {
+                        const GPUExecutionParameters gpu_execution_parameters{
+                            llvm_cuda_grid_dim_x,
+                            llvm_cuda_grid_dim_y,
+                            llvm_cuda_grid_dim_z,
+                            llvm_cuda_block_dim_x,
+                            llvm_cuda_block_dim_y,
+                            llvm_cuda_block_dim_z};
+                        benchmark::LLVMBenchmark benchmark(visitor,
+                                                           modfile,
+                                                           output_dir,
+                                                           shared_lib_paths,
+                                                           num_experiments,
+                                                           instance_size,
+                                                           platform,
+                                                           llvm_opt_level_ir,
+                                                           llvm_opt_level_codegen,
+                                                           gpu_execution_parameters);
+                        benchmark.run(ast);
+                    } else {
+                        benchmark::LLVMBenchmark benchmark(visitor,
+                                                           modfile,
+                                                           output_dir,
+                                                           shared_lib_paths,
+                                                           num_experiments,
+                                                           instance_size,
+                                                           platform,
+                                                           llvm_opt_level_ir,
+                                                           llvm_opt_level_codegen);
+                        benchmark.run(ast);
+                    }
```

Comment:
`nitpick`: can above duplication be removed as `platform` can be used to decide if passed `gpu_execution_parameters` can be ignored?

Reply:
I guess ideally  we would want to have 3 classes: Platform, GPUPlatform <: Platform, CPUPlatform <: Platform and all associated parameters can be passed there. Then in the code we can have a check if the platform is a GPU, say, cast it, and query grid/block dimensions.

I guess it is some work though, so having an extra optional argument to `Platform` constructor will work, as Pramod said :)